### ISSUE TITLE
Differences between Blinds with Fixed and Movable Slats 

### ISSUE THREAD

**Issue by amirroth** [NREL/EnergyPlus#10640]
In the course of working on a refactor, I stumbled upon an issue. This issue is illustrated quite clearly by two existing testfiles, `PurchAirWindowBlind.idf` and `_PurchAirWindowBlind2.idf`. Both use identical blinds, but the first uses fixed slats and the second uses movable ones. Look at the construction reports. First line is from `PAWB` (fixed slats) and the second is from `PAWB2` (movable slats). Both of these are evaluated with the slats at 45 degrees and normal solar incidence, i.e., Phi is 0. 

```cpp
blind1out.eio: WindowConstruction,WIN-CON-SINGLEPANE WITH INTERIOR BLIND,6,2,VerySmooth,5.894,5.894,1.000,0.588,0.196,0.132
blind2out.eio: WindowConstruction,WIN-CON-SINGLEPANE WITH INTERIOR BLIND,6,2,VerySmooth,5.894,5.894,1.000,0.577,0.192,0.129
```

Notice that the last three fields (which are SHGC, solar transmittance, and visible transmittance) are different. Why? Well, these values are computed differently for blinds with fixed and variable slats. For fixed-slat blinds they are computed at the slat angle (here 45 degrees). For movable-slat blinds, they are pre-computed at 10 degree increments and then interpolated (here, interpolated from values computed at 40 and 50 degrees). 

Now, there appears to be a peak for these values at the 45 degree slat angle, and when you sample at 10 degree increments you miss that peak. All of these values are complex expressions of other quantities which have monotonic behavior. Some are monotonic one way, some are monotonic the other way so that there is a peak at a or around 45 degrees. It doesn't really make sense to me that you would miss this peak by interpolating monotonic values, but that is the observed behavior in the code. Does this make sense to anyone? Is this legitimate or is there a problem with the implementation or a problem with my understanding of the results? 

Anyway, changing the slat angle sampling to 5 degree increments rather than 10 degrees makes this discrepancy go away, although of course it does introduce some diffs in test files with movable-slat blinds.

**Comment by mjwitte** [comment]
My guess is that the diffuse transmittance is higher for 45 than for 40, while the beam transmittance is lower (at normal incidence).
Can you see these component values in the nominal SHGC and transmittance calcs? Is it interpolating on the component results (beam and diffuse) or the total results?

**Comment by mjwitte** [comment]
10 degree sampling seems rather coarse, esp if these are one-time up front calculations. I might change that to 5 and check that in before the refactor. @vidanovic any thoughts on this?

**Comment by amirroth** [comment]
> 10 degree sampling seems rather coarse, esp if these are one-time up front calculations. I might change that to 5 and check that in before the refactor. @vidanovic any thoughts on this?

That's exactly what I did. I have a separate branch called `InterpSlatAng` that isolates this issue.

**Comment by amirroth** [comment]
> My guess is that the diffuse transmittance is higher for 45 than for 40, while the beam transmittance is lower (at normal incidence). Can you see these component values in the nominal SHGC and transmittance calcs? Is it interpolating on the component results (beam and diffuse) or the total results?

It's interpolating the component values and calculating the composites. I think the problem is that the behavior around 45 degrees is both non-linear and asymmetric, e.g., quantities change more from 45 to 50 degrees than they do from 40 to 45 degrees and vice versa. For instance, beam-to-diffuse transmittance is 0.149 at 45 degrees, 0.082 at 40 degrees, and 0.229 at 50 degrees. Relatively speaking, there is about a 15% greater difference between 45 and 50 than there is between 40 and 45.

**Comment by vidanovic** [comment]
> In the course of working on a refactor, I stumbled upon an issue. This issue is illustrated quite clearly by two existing testfiles, `PurchAirWindowBlind.idf` and `_PurchAirWindowBlind2.idf`. Both use identical blinds, but the first uses fixed slats and the second uses movable ones. Look at the construction reports. First line is from `PAWB` (fixed slats) and the second is from `PAWB2` (movable slats). Both of these are evaluated with the slats at 45 degrees and normal solar incidence, i.e., Phi is 0.
> 
> ```c++
> blind1out.eio: WindowConstruction,WIN-CON-SINGLEPANE WITH INTERIOR BLIND,6,2,VerySmooth,5.894,5.894,1.000,0.588,0.196,0.132
> blind2out.eio: WindowConstruction,WIN-CON-SINGLEPANE WITH INTERIOR BLIND,6,2,VerySmooth,5.894,5.894,1.000,0.577,0.192,0.129
> ```
> 
> Notice that the last three fields (which are SHGC, solar transmittance, and visible transmittance) are different. Why? Well, these values are computed differently for blinds with fixed and variable slats. For fixed-slat blinds they are computed at the slat angle (here 45 degrees). For movable-slat blinds, they are pre-computed at 10 degree increments and then interpolated (here, interpolated from values computed at 40 and 50 degrees).
> 
> Now, there appears to be a peak for these values at the 45 degree slat angle, and when you sample at 10 degree increments you miss that peak. All of these values are complex expressions of other quantities which have monotonic behavior. Some are monotonic one way, some are monotonic the other way so that there is a peak at a or around 45 degrees. It doesn't really make sense to me that you would miss this peak by interpolating monotonic values, but that is the observed behavior in the code. Does this make sense to anyone? Is this legitimate or is there a problem with the implementation or a problem with my understanding of the results?
> 
> Anyway, changing the slat angle sampling to 5 degree increments rather than 10 degrees makes this discrepancy go away, although of course it does introduce some diffs in test files with movable-slat blinds.

I would guess someone picked 10 degrees increment because it was considered to most optimal solution (speed and memory wise). I find extremely strange that pick is exactly at 45 degrees slat angle. How hard is to (just for test purposes) put increment at 1 degree and see where the pick is? I am assuming these are calculated with normal incidence but not running some kind of average over the some period since that would make calculations for non normal incidence, although I would still find hard to believe that incoming Sun direction is exactly at 45 degrees.

**Comment by amirroth** [comment]
> I would guess someone picked 10 degrees increment because it was considered to most optimal solution (speed and memory wise). I find extremely strange that pick is exactly at 45 degrees slat angle. How hard is to (just for test purposes) put increment at 1 degree and see where the pick is? I am assuming these are calculated with normal incidence but not running some kind of average over the some period since that would make calculations for non normal incidence, although I would still find hard to believe that incoming Sun direction is exactly at 45 degrees.

The 45 degrees is just "standard conditions" for the construction reports. And apparently the default angle for fixed-slat blinds. For annual simulation, the angle is going to be whatever it is set to be. 

Changing the sampling resolution to 1 degree is trivial. There is no code right now to find the peak SHGC/SolTrans/VisTran (the construction report is for 45 degrees, it doesn't search for peak values), but I could add that easily enough if it helps clarify what is happening here and what the right resolution is.

**Comment by vidanovic** [comment]
Maybe changing sampling to miss 45 degrees would say more. The above results might differ because in case 1 it calls venetian blind optical routines to calculate the data for the 45 degrees while in second case it interpolates between values calculated at 40 and 50?

**Comment by amirroth** [comment]
> Maybe changing sampling to miss 45 degrees would say more. The above results might differ because in case 1 it calls venetian blind optical routines to calculate the data for the 45 degrees while in second case it interpolates between values calculated at 40 and 50?

Yes. That is what's happening. Just trying to decide what to do about it.

**Comment by vidanovic** [comment]
I would propose add more interpolation points if that is not going to hamper speed too much. Maybe try one degree increment.

**Comment by amirroth** [comment]
Alright, so I printed out the range of TVisNorm [0..90] at one degree increments for three different calculations:

- Sampling at 1 degree increments (essentially no interpolation at all)
- Sampling at 5 degree increments and interpolating
- Sampling at 10 degree increments and interpolating (the current implementation)

Most of the shenanigans take place between 40 and 50 degrees where the behavior is most non-linear. 

- With sampling at 10 degree increments, there is a 10-15% relative under-calculation in this range. Although that falls off rapidly and the errors are negligible in the range of 20-40 and 50-70 and overcalculations of about 5-10% in the 0-20 and 70-90 degree ranges.
- With sampling at 5 degree increments, there is a 10% relative under-calculation in the 40-45 degree range, but almost no error in the 45-50 degree range! Overcalculations in the 0-20 and 70-90 degree ranges are about 5%. 

As far as recommendations, we should go with at least 5 degree sampling and maybe 1 degree. There will be somewhat of a memory explosion for daylighting and solar shading calculations if we do this, but the refactor that uncovered this issue will eliminate those (in this refactor, `SlatAngle` is only a dimension in one-time `Material` and `Construction` level calculations, not in individual window and daylighting reference point calculations as they are now, for those there is only the current slat angle for which all values were interpolated the last time that angle changed).

Vote, 5 degree or 1 degree?

**Comment by vidanovic** [comment]
I would say keep it at 1 degree (my vote).

**Comment by amirroth** [comment]
For the record, it's currently at 10 degrees.

**Comment by mjwitte** [comment]
Given the 10% interpolation error even with 5 degree, I'm leaning toward 1 degree increments.
These extra calcs are only done per unique shaded construction, correct? e.g. If there are 100 windows all of the same construction with the same windowblind material, this is only done once.

**Comment by vidanovic** [comment]
> Given the 10% interpolation error even with 5 degree, I'm leaning toward 1 degree increments. These extra calcs are only done per unique shaded construction, correct? e.g. If there are 100 windows all of the same construction with the same windowblind material, this is only done once.

Yes, this should be true (it was the last time I looked at it). No reason to keep recalculating optical properties.

**Comment by amirroth** [comment]
Ok, I will do 1 degree and we'll see what the performance implications are. Note, the 10% difference is relative not absolute. It's the difference between 0.15 and 0.135, not the difference between 0.15 and 0.05.

**Comment by mjwitte** [comment]
@amirroth #10646 closes this issue, correct?

**Comment by mjwitte** [comment]
Closed via #10646
